src/Init/System/IO.lean

```diff
@@ -32,106 +42,202 @@ instance : MonadExceptOf ε (EIO ε) := inferInstanceAs (MonadExceptOf ε (EStat
+/--
+Runs a `BaseIO` action, which cannot throw an exception, in any other `EIO` monad.
+
+This function is usually used implicitly via [automatic monadic
+lifting](lean-manual://section/lifting-monads) rather being than called explicitly.
+-/
 @[always_inline, inline]
 def BaseIO.toEIO (act : BaseIO α) : EIO ε α :=
   fun s => match act s with
   | EStateM.Result.ok a s => EStateM.Result.ok a s
 
 instance : MonadLift BaseIO (EIO ε) := ⟨BaseIO.toEIO⟩
 
+/--
+Converts an `EIO ε` action that might throw an exception of type `ε` into an exception-free `BaseIO`
+action that returns an `Except` value.
+-/
 @[always_inline, inline]
 def EIO.toBaseIO (act : EIO ε α) : BaseIO (Except ε α) :=
   fun s => match act s with
   | EStateM.Result.ok a s     => EStateM.Result.ok (Except.ok a) s
   | EStateM.Result.error ex s => EStateM.Result.ok (Except.error ex) s
 
+/--
+Handles any exception that might be thrown by an `EIO ε` action, transforming it into an
+exception-free `BaseIO` action.
+-/
 @[always_inline, inline]
 def EIO.catchExceptions (act : EIO ε α) (h : ε → BaseIO α) : BaseIO α :=
   fun s => match act s with
   | EStateM.Result.ok a s     => EStateM.Result.ok a s
   | EStateM.Result.error ex s => h ex s
 
+/--
+Converts an `Except ε` action into an `EIO ε` action.
+
+If the `Except ε` action throws an exception, then the resulting `EIO ε` action throws the same
+exception. Otherwise, the value is returned.
+-/
 def EIO.ofExcept (e : Except ε α) : EIO ε α :=
   match e with
   | Except.ok a    => pure a
   | Except.error e => throw e
 
 open IO (Error) in
+/--
+A monad that supports arbitrary side effects and throwing exceptions of type `IO.Error`.
+-/
 abbrev IO : Type → Type := EIO Error
 
+/--
+Runs a `BaseIO` action, which cannot throw an exception, as an `IO` action.
+
+This function is usually used implicitly via [automatic monadic
+lifting](lean-manual://section/lifting-monads) rather than being called explicitly.
+-/
 @[inline] def BaseIO.toIO (act : BaseIO α) : IO α :=
   act
 
+/--
+Converts an `EIO ε` action into an `IO` action by translating any exceptions that it throws into
+`IO.Error`s using `f`.
+-/
 @[inline] def EIO.toIO (f : ε → IO.Error) (act : EIO ε α) : IO α :=
   act.adaptExcept f
 
+/--
+Converts an `EIO ε` action that might throw an exception of type `ε` into an exception-free `IO`
+action that returns an `Except` value.
+-/
 @[inline] def EIO.toIO' (act : EIO ε α) : IO (Except ε α) :=
   act.toBaseIO
 
+/--
+Runs an `IO` action in some other `EIO` monad, using `f` to translate `IO` exceptions.
+-/
 @[inline] def IO.toEIO (f : IO.Error → ε) (act : IO α) : EIO ε α :=
   act.adaptExcept f
 
 /- After we inline `EState.run'`, the closed term `((), ())` is generated, where the second `()`
    represents the "initial world". We don't want to cache this closed term. So, we disable
    the "extract closed terms" optimization. -/
 set_option compiler.extract_closed false in
+/--
+Executes arbitrary side effects in a pure context.
```

Comment:
This function needs a bigger docstring to explain the dangers of using it, much like the one on [Haskell's unsafePerformIO](https://hackage.haskell.org/package/base-4.21.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO). In particular, it should mention that calls to this function may be duplicated or eliminated because the resulting value is treated as if it was a pure expression. Also some variant on this remark:

> WARNING: If you're looking for "a way to get a String from an 'IO String'", then [unsafePerformIO](https://hackage.haskell.org/package/base-4.21.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO) is not the way to go. Learn about do-notation and the <- syntax element before you proceed.

Reply:
Thanks, this is very helpful!